pgsql/pc_editor.c

```diff
@@ -21,32 +21,46 @@ Datum pcpatch_setschema(PG_FUNCTION_ARGS)
 	int32 mode = PG_GETARG_INT32(3);
 	PCSCHEMA *schema = pc_schema_from_pcid(serpa->pcid, fcinfo);
 	PCSCHEMA *new_schema = pc_schema_from_pcid(pcid, fcinfo);
-	PCPATCH *patch = pc_patch_deserialize(serpa, schema);
-	bool reinterpret;
-	float8 defaultvalue;
+	PCPATCH *patch;
+	float8 defaultvalue = 0.0;
 
-	if ( ! patch )
-		PG_RETURN_NULL();
-
-	switch ( mode )
+	if ( mode == 0 )
 	{
-	case 0:
+		bool reinterpret;
+		if ( pc_schema_similar(schema, new_schema) )
+		{
+			// fast path!
+			serpatch = palloc(serpa->size);
+			if ( ! serpatch )
+				PG_RETURN_NULL();
+			memcpy(serpatch, serpa, serpa->size);
+			serpatch->pcid = pcid;
+			PG_RETURN_POINTER(serpatch);
```

Comment:
See a775092 for the tests. The "pcid 1 to 3" fails without the commit that deals with the case of similar patches with different compressions (96abb94).

This is `regression.diffs`:

```
*** /home/elemoine/src/pointcloud/pgsql/expected/pointcloud.out 2017-04-04 11:02:31.559558781 +0200
--- /home/elemoine/src/pointcloud/pgsql/results/pointcloud.out  2017-04-04 11:02:52.327609106 +0200
***************
*** 619,627 ****
  SELECT
    PC_AsText(PC_SetPCId(p, 3)) t, PC_Summary(PC_SetPCId(p, 3))::json->'compr' c
  FROM ( SELECT PC_Patch(PC_MakePoint(1, ARRAY[-1,0,4862413,1])) p ) foo;
!                     t                    |       c
! -----------------------------------------+---------------
!  {"pcid":3,"pts":[[-1,0,4.86241e+06,1]]} | "dimensional"
  (1 row)

  -- test PC_SetPCId
--- 619,627 ----
  SELECT
    PC_AsText(PC_SetPCId(p, 3)) t, PC_Summary(PC_SetPCId(p, 3))::json->'compr' c
  FROM ( SELECT PC_Patch(PC_MakePoint(1, ARRAY[-1,0,4862413,1])) p ) foo;
!                     t                    |   c
! -----------------------------------------+--------
!  {"pcid":3,"pts":[[-1,0,4.86241e+06,1]]} | "none"
  (1 row)

  -- test PC_SetPCId

======================================================================
```